pkg/component/gardener/apiserver/deployment.go

```diff
@@ -147,5 +147,21 @@ func (g *gardenerAPIServer) deployment(
 
 	utilruntime.Must(gardenerutils.InjectGenericKubeconfig(deployment, secretGenericTokenKubeconfig.Name, secretVirtualGardenAccess.Secret.Name))
 	utilruntime.Must(references.InjectAnnotations(deployment))
+
+	// Preserve the Deployment resources during the HPA -> VPAAndHPA transition.
+	//
+	// The gardener-apiserver is deployed via ManagedResource. In the general case, GRM preserves the Deployment resources when the resource is being scaled by HVPA.
+	// However, druing the HPA -> VPAAndHPA transition the flow is as follows:
+	// 1. The HVPA resource is deleted.
+	// 2. The Deployment is updated.
+	// 3. The new HPA and VPA resources are created.
+	//
+	// Step 2. is the one that reverts the Deployment resources to the initial ones as during that time there is no existing HVPA object.
+	//
+	// TODO(ialidzhikov): Drop this if statement when the VPAAndHPA is the only one supported autoscaling mode.
+	if g.values.Autoscaling.Mode == apiserver.AutoscalingModeVPAAndHPA {
+		metav1.SetMetaDataAnnotation(&deployment.ObjectMeta, resourcesv1alpha1.PreserveResources, "true")
```

Comment:
While we're doing this: do we also need to set `resourcesv1alpha1.PreserveReplicas` here?

Reply:
Hi @voelzmo ,

I tried to set this on Friday and I tried to explain the outcome in the `Special notes for your reviewer` section.
Even if GRM preserves the replicas when updating the Deployment, the replicas are always set by the highavailabilityconfig webhook of the GRM. See https://github.com/gardener/gardener/blob/db182529c3c1badd987f7e5b20bdcc71f33f8e0a/pkg/resourcemanager/webhook/highavailabilityconfig/handler.go#L313-L318
During the migration HVPA -> VPAAndHPA, HVPA will be deleted first, then the Deployment gets updated. Hence, `isHorizontallyScaled` will be false and the HA config webhook of GRM will mutate the replicas.

tl;dr: According to my testing setting `resourcesv1alpha1.PreserveReplicas` does not have effect.

Reply:
Ah, thanks for explaining it in a bit more detail. I had read this section, but didn't really understand that the annotation doesn't apply in this case. I get now that this is due to the flow of deleting the HPA object, updating Deployment (and thereby HA config webhook setting replicas to 2) and only then creating a new HPA object.
If we really want, we could avoid this by setting `high-availability-config.resources.gardener.cloud/replicas: 3`, not sure if this is really necessary as this will anyways happen when the HPA gets created with `minReplicas: 3`?

Reply:
Hm, IIUC the proposal is to make use of `high-availability-config.resources.gardener.cloud/replicas: 3`. According to https://github.com/gardener/gardener/blob/450529c84b99ddc36a3ec43abb1e530839fc18a5/docs/concepts/resource-manager.md#L964-L967, the HA webhook of GRM does NOT mutate the replicas
> when the resource is scaled horizontally by `HorizontalPodAutoscaler` or `Hvpa`, and the current replica count is higher than what was computed above.

Hence, `high-availability-config.resources.gardener.cloud/replicas: 3` will be used only during the migration from HVPA to HPAAndVPA autoscaling mode. Then, after the HPA is created for the gardener-apiserver, HPA will control the replicas count, the HA webhook won't overwrite the replicas count and the `high-availability-config.resources.gardener.cloud/replicas` annotation on the Deployment won't be respected.

It looks hacky but I can test and if it works, it improves a little bit the transition.

Reply:
I played with the above proposal in local setup. For some reason that I didn't fully investigate, even with `high-availability-config.resources.gardener.cloud/replicas: 3` I see minDomains flapping between 2 <-> 3 during the migration from HVPA mode with currentReplicas=4,minReplicas=4,maxReplicas=4 (just for testing I adapted the HVPA to run with 4 replicas for a more real world testing) and VPAAndMode minReplicas=3,maxReplicas=6.

It is not super clear proposal but rather relying on implicit contracts. And at the end it does not avoid the new ReplicaSet creation (due this minDomains explained in the above sentence). So, I would rather leave it as it is and accept the 4 -> 2 scale down, then HPA scale up to 3 replicas and then the new ReplicaSet due to the minDomains change. At the end, we will have at least 2 running replicas during the migration so it is not super bad - just fluctuation/noise during the migration which should stabilize with the new ReplicaSet.